src/table/src/predicate.rs

```diff
@@ -12,66 +12,92 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::sync::Arc;
+
 use common_query::logical_plan::{DfExpr, Expr};
 use common_telemetry::{error, warn};
 use common_time::range::TimestampRange;
+use common_time::timestamp::TimeUnit;
 use common_time::Timestamp;
 use datafusion::parquet::file::metadata::RowGroupMetaData;
 use datafusion::physical_optimizer::pruning::PruningPredicate;
 use datafusion_common::ToDFSchema;
 use datafusion_expr::expr::InList;
 use datafusion_expr::{Between, BinaryExpr, Operator};
-use datafusion_physical_expr::create_physical_expr;
 use datafusion_physical_expr::execution_props::ExecutionProps;
+use datafusion_physical_expr::{create_physical_expr, PhysicalExpr};
 use datatypes::schema::SchemaRef;
 use datatypes::value::scalar_value_to_timestamp;
+use snafu::ResultExt;
 
+use crate::error;
 use crate::predicate::stats::RowGroupPruningStatistics;
 
 mod stats;
 
-#[derive(Default, Clone)]
+#[derive(Clone)]
 pub struct Predicate {
-    exprs: Vec<Expr>,
+    /// The schema of underlying storage.
+    schema: SchemaRef,
+    /// Physical expressions of this predicate.
+    exprs: Vec<Arc<dyn PhysicalExpr>>,
 }
 
 impl Predicate {
-    pub fn new(exprs: Vec<Expr>) -> Self {
-        Self { exprs }
+    /// Creates a new `Predicate` by converting logical exprs to physical exprs that can be
+    /// evaluated against record batches.
+    /// Returns error when failed to converting exprs.
+    pub fn try_new(exprs: Vec<Expr>, schema: SchemaRef) -> error::Result<Self> {
+        let arrow_schema = schema.arrow_schema();
+        let df_schema = arrow_schema
+            .clone()
+            .to_dfschema_ref()
+            .context(error::DatafusionSnafu)?;
+
+        // `execution_props` provides variables required by evaluation.
+        let execution_props = &ExecutionProps::new();
```

Comment:
I think `execution_props` should be retrieved from the query engine to respect the query engine settings.

https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html#method.execution_props

https://github.com/GreptimeTeam/greptimedb/blob/f287d3115b5ea38f72c9267d8437ecb6067ff57c/src/query/src/query_engine/state.rs#L57

Reply:
I did not use `ExectionProps` from `SesstionState` in that:
- We cannot pass the session state down to the table scan stage.
- The `SessionState::execution_props` field is [initialize to `ExecutionProps::new()`](https://docs.rs/datafusion/25.0.0/src/datafusion/execution/context.rs.html#1488) and we never register any variable provider to `SessionState`, thus it's the same as `ExecutionProps::new()` here.

Reply:
Yep, but when we start to register some variables to the session state, we may forget here. 
Maybe we can add a TODO or a global variable in query crate.